core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -99,95 +94,104 @@ func NewLogBuffer(lggr logger.Logger, lookback, blockRate, logLimit uint32) LogB
-	defer b.enqueuedBlockLock.Unlock()
-
-	if uid == nil {
-		return
+	if len(reorgBlocks) > 0 {
+		b.evictReorgdLogs(reorgBlocks)
 	}
 
-	for blockNumber := range uniqueBlocks {
-		if blockNumbers, ok := b.enqueuedBlocks[blockNumber]; ok {
-			if upkeepBlockInstances, ok := blockNumbers[uid.String()]; ok {
-				blockNumbers[uid.String()] = upkeepBlockInstances + 1
-				b.lggr.Debugw("enqueuing logs again for a previously seen block for this upkeep", "blockNumber", blockNumber, "numberOfEnqueues", b.enqueuedBlocks[blockNumber], "upkeepID", uid.String())
-			} else {
-				blockNumbers[uid.String()] = 1
-			}
-			b.enqueuedBlocks[blockNumber] = blockNumbers
-		} else {
-			b.enqueuedBlocks[blockNumber] = map[string]int{
-				uid.String(): 1,
+	return latest
+}
+
+func (b *logBuffer) evictReorgdLogs(reorgBlocks map[int64]bool) {
+	for blockNumber := range reorgBlocks {
+		start, _ := getBlockWindow(blockNumber, int(b.opts.blockRate.Load()))
+		for _, queue := range b.queues {
+			if _, ok := queue.logs[blockNumber]; ok {
+				queue.logs[blockNumber] = []logpoller.Log{}
+				queue.dequeued[start] = 0
 			}
 		}
 	}
 }
 
 // Dequeue greedly pulls logs from the buffers.
 // Returns logs and the number of remaining logs in the buffer.
-func (b *logBuffer) Dequeue(block int64, blockRate, upkeepLimit, maxResults int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+func (b *logBuffer) Dequeue(startWindowBlock int64, maxResults int, bestEffort bool) ([]BufferedLog, int) {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	start, end := getBlockWindow(block, blockRate)
-	return b.dequeue(start, end, upkeepLimit, maxResults, upkeepSelector)
+	return b.dequeue(startWindowBlock, maxResults, bestEffort)
 }
 
-// dequeue pulls logs from the buffers, depends the given selector (upkeepSelector),
-// in block range [start,end] with minimum number of results per upkeep (upkeepLimit)
-// and the maximum number of results (capacity).
+// dequeue pulls logs from the buffers, in block range [start,end] with minimum number
+// of results per upkeep (upkeepLimit) and the maximum number of results (capacity).
+// If operating under minimum dequeue, upkeeps are skipped when the minimum number
+// of logs have been dequeued for that upkeep.
 // Returns logs and the number of remaining logs in the buffer for the given range and selector.
 // NOTE: this method is not thread safe and should be called within a lock.
-func (b *logBuffer) dequeue(start, end int64, upkeepLimit, capacity int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+func (b *logBuffer) dequeue(start int64, capacity int, minimumDequeue bool) ([]BufferedLog, int) {
 	var result []BufferedLog
 	var remainingLogs int
-	for _, q := range b.queues {
-		if !upkeepSelector(q.id) {
-			// if the upkeep is not selected, skip it
+	minimumDequeueMet := 0
+
+	logLimit := int(b.opts.logLimit.Load())
+	end := start + int64(b.opts.blockRate.Load())
+
+	if !minimumDequeue {
+		logLimit = capacity
+	}
```

Comment:
nit. would be clearer to not override logLimit as that has a fixed meaning of the config we get. 
Instead we can use capacity directly in `q.dequeue`